crates/symbolicator/src/services/shared_cache.rs

```diff
@@ -0,0 +1,346 @@
+    fn gcs_bucket_key(&self) -> String {
+        // All our paths should be UTF-8, we don't construct non-UTF-8 paths.
+        match self.relative_path().to_str() {
+            Some(s) => s.to_owned(),
+            None => {
+                log::error!(
+                    "Non UTF-8 path in SharedCacheKey: {}",
+                    self.relative_path().display()
+                );
+                self.relative_path().to_string_lossy().into_owned()
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+enum SharedCacheBackend {
+    Gcs(GcsState),
+    Fs(FilesystemSharedCacheConfig),
+}
+
+/// A shared cache service.\
+///
+/// For simplicity in the rest of the application this service always exists, regardless of
+/// whether it is configured or not.  If it is not configured this becomes a no-op.
+#[derive(Debug)]
+pub struct SharedCacheService {
+    backend: Option<SharedCacheBackend>,
+}
+
+impl SharedCacheService {
+    pub fn new(config: Option<SharedCacheConfig>) -> Self {
+        // Yes, it is wasteful to create a GcsState when the config does not use GCS.
+        // However currently we only have Filesystem as alternative and that is only used
+        // for testing purposes.
+        let backend = config.map(|cfg| match cfg {
+            SharedCacheConfig::Gcs(cfg) => SharedCacheBackend::Gcs(GcsState::new(cfg)),
+            SharedCacheConfig::Fs(cfg) => SharedCacheBackend::Fs(cfg),
+        });
+        Self { backend }
+    }
+
+    /// Returns the name of the backend configured.
+    fn backend_name(&self) -> &str {
+        match self.backend {
+            Some(SharedCacheBackend::Fs(_)) => "filesystem",
+            Some(SharedCacheBackend::Gcs(_)) => "GCS",
+            None => "<not-configured>",
+        }
+    }
+
+    /// Retrieve a file from the shared cache.
+    ///
+    /// `cache_path` is the relative path on the shared cache.  `destination` is where to
+    /// write the file locally.
+    ///
+    /// Errors are transparently hidden, either a cache item is available or it is not.
+    pub async fn fetch<W>(&self, key: &SharedCacheKey, writer: &mut W) -> bool
+    where
+        W: tokio::io::AsyncWrite + Unpin,
+    {
+        let res = match self.backend {
+            Some(SharedCacheBackend::Gcs(ref state)) => state.fetch(key, writer).await,
+            Some(SharedCacheBackend::Fs(ref cfg)) => cfg.fetch(key, writer).await,
+            None => return false,
+        };
+        match res {
+            Ok(_) => {
+                metric!(counter(&format!("services.shared_cache.{}", key.name)) += 1, "hit" => "true");
+                true
+            }
+            Err(err) => {
+                metric!(counter(&format!("services.shared_cache.{}", key.name)) += 1, "hit" => "false");
+                log::error!(
+                    "Error fetching from {} shared cache: {}",
+                    self.backend_name(),
+                    LogError(&*err)
+                );
```

Comment:
I haven’t figured out from the code if we are raising errors for each "not found"? We shouldn’t, as that is totally an expected outcome.

Reply:
ugh, yes.  good catch